### PR TITLE
sql: fix possible nil pointer in crdb_internal.cluster_locks

### LINKED ISSUE #113935: sql: TestLogic_cluster_locks fails with nil pointer dereference
Found in unrelated PR https://github.com/cockroachdb/cockroach/pull/113527

TC: https://teamcity.cockroachdb.com/viewLog.html?buildId=12535137&buildTypeId=Cockroach_BazelEssentialCi

```
=== RUN   TestLogic_cluster_locks
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/0dbccfa3ae257c24fce6e73a3689e99e/logTestLogic_cluster_locks2318709849
    test_log_scope.go:81: use -show-logs to present logs inline
I231106 17:43:45.320848 1 (gostd) rand.go:250  [-] 1  random seed: 7705222062379041669
    logic.go:2966: let $root_session = 179519a40557a6810000000000000001
    logic.go:2966: let $testuser_session = 179519a41d7b5d1e0000000000000001
    logic.go:2966: let $txn1 = 0009ab4e-7fda-489c-998a-db0d93e58c2a
    logic.go:2966: let $txn2 = 589e476b-4945-4443-ae18-77089d4e6f45
    logic.go:2966: let $r1 = 67
    logic.go:2966: let $r2 = 68
    logic.go:2966: let $r3 = 69
    logic.go:2966: let $txn3 = 392f7976-139c-4294-90d4-3a1b78d90021
    logic.go:2966: let $txn4 = a049e81a-bc6a-4a5f-ada0-48c34dd745f0
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x358a415]

goroutine 3662 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0042d7730?, {0x6bb85d8, 0xc007416b40})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x65
panic({0x4d7f680?, 0x9c61710?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/cockroach/pkg/sql.populateClusterLocksWithFilter.genClusterLocksGenerator.func1.6()
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8022 +0x735
github.com/cockroachdb/cockroach/pkg/sql.populateClusterLocksWithFilter({0x6bb85d8?, 0xc007416b40?}, 0x6b74fa0?, {0x6c58218?, 0xc004c92bb0?}, 0xc00237d4d0, {0x0, 0x0, 0xc003bcc9e0, 0x0})
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8058 +0xac
github.com/cockroachdb/cockroach/pkg/sql.init.genPopulateClusterLocksWithIndex.func117({0x6bb85d8, 0xc007416b40}, {0x6c18c20?, 0xc00726cc58}, 0x0?, {0x6c58218, 0xc004c92bb0}, 0x4?)
	github.com/cockroachdb/cockroach/pkg/sql/crdb_internal.go:8040 +0x11a
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).getPlanInfo.func1.(*virtualDefEntry).makeConstrainedRowsGenerator.func2({0x6bb85d8, 0xc007416b40}, {0x6b74fa0, 0xc00125e920})
	github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:823 +0x558
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3({0x6bb85d8, 0xc007416b40})
	github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:127 +0x102
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x13a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 2725
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x415
```

This is the line of code in crdb_internal.go that panics:
```
				tree.NewDString(tree.IsolationLevelFromKVTxnIsolationLevel(curLock.LockHolder.IsoLevel).String()), /* isolation_level */
```

I have not been able to repro this, even with the random seed.

Jira issue: CRDB-33263

### PR DESCRIPTION
**Pull request by yuzefovich**
In fa33c28abd95f5c0c586484b99dedc5597356a94 we introduced a regression that could lead to a nil pointer when `curLock.LockHolder` is nil (which is valid). This is now fixed.

No release note since we haven't had a stable release with the bug yet.

Fixes: #113935.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
I spent some time trying to come up with a regression test, but I don't know quite understand the conditions required. @nvanbenschoten do you have some tips for that?

**Review by nvb — COMMENTED** [review]
A regression test for the crash might be difficult to come up with. We need a non-transactional waiter in a lock wait-queue. Something like a backup stuck on a high priority writing transaction would do the trick, though that might be tricky to achieve in a logictest.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-:-NjP8xQO40Q0Dz5117G0:b-vje8f8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @yuzefovich)

___
*[`pkg/sql/crdb_internal.go` line 7950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjP8t5Q8BdygK11JZrf:-NjP8t5Q8BdygK11JZrg:b2yh0py) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbd710ddf25e211b09e93395bebbcb4995b66b1/pkg/sql/crdb_internal.go#L7950)):*
> ```Go
> 					txnIDDatum = tree.NewDUuid(tree.DUuid{UUID: waiter.WaitingTxn.ID})
> 					tsDatum = eval.TimestampToInexactDTimestamp(waiter.WaitingTxn.WriteTimestamp)
> 					isolationLevelDatum = tree.NewDString(tree.IsolationLevelFromKVTxnIsolationLevel(waiter.WaitingTxn.IsoLevel).String())
> ```

It's unfortunate that none of our testing caught this bug. In fact, none of the tests use other isolation levels.

Can we throw some Read Committed and Snapshot transactions into `pkg/sql/logictest/testdata/logic_test/cluster_locks` to improve the test coverage?

To do so, I think we'll need to set `sql.txn.read_committed_isolation.enabled` and `sql.txn.snapshot_isolation.enabled`to true.

We'll also want to keep in mind that non-locking reads don't block under these isolation levels, so we'll have better luck with `SELECT FOR UPDATE`.

One possible test would be:
```sql
BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;  SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;    SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;  SELECT * FROM t WHERE k = 'a' FOR UPDATE;
```
Basically, we just need to build up a wait-queue on a single key with non-serializable waiters.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @nvanbenschoten)

___
*[`pkg/sql/crdb_internal.go` line 7950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjP8t5Q8BdygK11JZrf:-NjPIvG_AqdIZ1rAEGyh:bai06v3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbd710ddf25e211b09e93395bebbcb4995b66b1/pkg/sql/crdb_internal.go#L7950)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's unfortunate that none of our testing caught this bug. In fact, none of the tests use other isolation levels.

Can we throw some Read Committed and Snapshot transactions into `pkg/sql/logictest/testdata/logic_test/cluster_locks` to improve the test coverage?

To do so, I think we'll need to set `sql.txn.read_committed_isolation.enabled` and `sql.txn.snapshot_isolation.enabled`to true.

We'll also want to keep in mind that non-locking reads don't block under these isolation levels, so we'll have better luck with `SELECT FOR UPDATE`.

One possible test would be:
```sql
BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;  SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;    SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; SELECT * FROM t WHERE k = 'a' FOR UPDATE;
BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;  SELECT * FROM t WHERE k = 'a' FOR UPDATE;
```
Basically, we just need to build up a wait-queue on a single key with non-serializable waiters.
</blockquote></details>

Thanks, done.

Relatedly, I noticed that we have a separate `cluster_locks_tenant` file, and I spent some time trying to unify them, but the `cluster_locks` table has different contents. Do you have an idea why that would be the case?

**Review by nvb — APPROVED** [review]
:lgtm: thanks for the quick fix!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-:-NjPPCehBLHwm47wGRUT:b2gl5d9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @annrpom and @yuzefovich)

___
*[`pkg/sql/crdb_internal.go` line 7950 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjP8t5Q8BdygK11JZrf:-NjPPHQSC-rDL5Bw7__Q:b-r6nzae) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cbd710ddf25e211b09e93395bebbcb4995b66b1/pkg/sql/crdb_internal.go#L7950)):*
> Do you have an idea why that would be the case?

No sorry, I don't know.
___
*[`pkg/sql/logictest/testdata/logic_test/cluster_locks` line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjPP4Yi8mCfogj2yaCR:-NjPP4Yi8mCfogj2yaCS:b-knqpr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4862a419529ce31675cf71d103cda020b97984f/pkg/sql/logictest/testdata/logic_test/cluster_locks#L375)):*
> ```
> 
> statement async iso2
> SELECT * FROM t WHERE k = 'a' FOR UPDATE;
> ```

Consider adding in a READ COMMITTED waiter as well.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @annrpom and @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/cluster_locks` line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjPP4Yi8mCfogj2yaCR:-NjPWRBw28zs462S50rG:b-pzvepk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4862a419529ce31675cf71d103cda020b97984f/pkg/sql/logictest/testdata/logic_test/cluster_locks#L375)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider adding in a READ COMMITTED waiter as well.
</blockquote></details>

Done. I think this requires adding certs for another test user, which I pulled out in a separate commit. Let's see whether it breaks the CI.

**Review by nvb — APPROVED** [review]
:lgtm:

> Reviewed 32 of 32 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-:-NjPWwGx3bi6Jfl7cL3J:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @annrpom)

**Review by annrpom — APPROVED** [review]
dang. i should've added better tests - i wonder why i didn't at the time :/

thank you for the fix and thank you all for catching this. the cluster_locks change lgtm

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114617)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[`pkg/sql/logictest/testdata/logic_test/cluster_locks` line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/114617#-NjPP4Yi8mCfogj2yaCR:-NjTc2rH2PhyGyP0Mi2B:b-xmpdhb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4862a419529ce31675cf71d103cda020b97984f/pkg/sql/logictest/testdata/logic_test/cluster_locks#L375)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Done. I think this requires adding certs for another test user, which I pulled out in a separate commit. Let's see whether it breaks the CI.
</blockquote></details>

The test which had 4 concurrent txns would occasionally get stuck, my guess is due to non-deterministic order in which waiters are woken up.  I filed #114660 to investigate that (Nathan, please feel free to close that issue if it doesn't seem worth investigating), but for this PR I split the scenario into two, to cover different isolation levels while only having exactly one waiter.

I also decided to keep the first commit in case we need `testuser3` in the future. Let's whether CI is happy now.

**Comment by yuzefovich** [comment]
TFTRs!

bors r+
